File: map-gen-script/src/map_generator/topologies/straight_line.py
# src/map_generator/topologies/straight_line.py

import random
import copy
from .base_topology import BaseTopology
from src.map_generator.models.path_info import PathInfo, Coord

# --- ĐẢM BẢO TÊN LỚP NÀY HOÀN TOÀN CHÍNH XÁC ---
class StraightLineTopology(BaseTopology):
    """
    Tạo ra một con đường thẳng dài trên một trục.
    
    Đây là một topology đa năng, được sử dụng cho nhiều chủ đề khác nhau
    như Vòng lặp For, Biến, và Vòng lặp While.
    """
    def generate_path_info_variants(self, grid_size: tuple, params: dict, max_variants: int) -> "Iterator[PathInfo]":
        """
        [REWRITTEN] Tạo biến thể bắt đầu từ params gốc, sau đó tăng dần độ dài.
        Điều này đảm bảo biến thể đầu tiên luôn khớp với curriculum.
        """
        base_length = int(params.get('path_length', 5))
        max_possible_length = min(grid_size[0] - 4, grid_size[2] - 4) # Trừ lề an toàn
        
        # 1. Luôn tạo biến thể gốc đầu tiên
        yield self.generate_path_info(grid_size, params)
        
        # 2. Tạo các biến thể tiếp theo bằng cách tăng dần độ dài
        for i in range(1, max_variants):
            new_length = base_length + i
            # Dừng lại nếu đường đi quá dài so với map
            if new_length > max_possible_length:
                break
                
            variant_params = copy.deepcopy(params)
            variant_params['path_length'] = new_length
            yield self.generate_path_info(grid_size, variant_params)
    
    def generate_path_info(self, grid_size: tuple, params: dict) -> PathInfo:
        """
        Tạo ra một đường thẳng với chiều dài và hướng ngẫu nhiên.

        Args:
            params (dict): Có thể chứa 'path_length' để tùy chỉnh chiều dài.

        Returns:
            PathInfo: Một đối tượng chứa thông tin về đường thẳng.
                      'path_coords' sẽ chứa tọa độ của các ô trên đường đi.
        """
        print("    LOG: Generating 'straight_line' topology...")
        
        # Logic sinh map gốc giờ sẽ đọc trực tiếp từ params đã được biến đổi
        path_length = int(params.get('path_length', 5))

        # Đảm bảo đường đi không quá dài so với map
        max_dim = max(grid_size[0], grid_size[2])
        if path_length >= max_dim - 2:
            path_length = max_dim - 3

        # Chọn một trục (X hoặc Z) và hướng (tiến hoặc lùi) ngẫu nhiên
        axis = random.choice(['x', 'z'])
        direction = random.choice([1, -1])
        
        # Tính toán vị trí bắt đầu an toàn
        total_cells_needed = path_length + 2
        
        y = 0
        start_pos_list = [0, y, 0]
        
        if axis == 'x':
            if direction == 1:
                start_x = random.randint(0, grid_size[0] - total_cells_needed)
            else:
                start_x = random.randint(total_cells_needed - 1, grid_size[0] - 1)
            start_pos_list[0] = start_x
            start_pos_list[2] = random.randint(0, grid_size[2] - 1)
        else: # axis == 'z'
            if direction == 1:
                start_z = random.randint(0, grid_size[2] - total_cells_needed)
            else:
                start_z = random.randint(total_cells_needed - 1, grid_size[2] - 1)
            start_pos_list[2] = start_z
            start_pos_list[0] = random.randint(0, grid_size[0] - 1)
        
        start_pos: Coord = tuple(start_pos_list)
        
        # Tạo ra danh sách các tọa độ trên đường đi
        # [SỬA LỖI] Bắt đầu đường đi với vị trí start_pos để đảm bảo có khối nền
        path_coords: list[Coord] = [start_pos]
        current_pos = list(start_pos)
        
        # Tạo các ô ở giữa
        for _ in range(path_length):
            if axis == 'x':
                current_pos[0] += direction
            else:
                current_pos[2] += direction
            path_coords.append(tuple(current_pos))
        
        target_pos = path_coords[-1]
        
        # [NEW] Tạo metadata cho Placers
        metadata = {
            "segment": path_coords,
            "path_length": path_length,
            "axis": axis,
            "direction": direction,
        }
        
        return PathInfo(
            start_pos=start_pos,
            target_pos=target_pos,
            path_coords=path_coords,
            placement_coords=path_coords, # [CHUẨN HÓA] Đảm bảo tất cả các ô đều có nền
            metadata=metadata
        )

File: packages/academic-map-generator/src/generator/topologies/StraightLine.ts
/**
 * StraightLine Topology (PORTED FROM PYTHON)
 * Creates a simple straight path on an axis.
 */

import { BaseTopology } from './BaseTopology';
import { IPathInfo, Coord } from '../types';

export class StraightLineTopology extends BaseTopology {
  *generatePathInfoVariants(
    gridSize: [number, number, number],
    params: Record<string, any>,
    maxVariants: number
  ): Generator<IPathInfo> {
    const baseLength = params.path_length || 5;
    const maxPossibleLength = Math.min(gridSize[0] - 4, gridSize[2] - 4);

    // 1. Yield base variant
    yield this.generatePathInfo(gridSize, params);

    // 2. Variants increasing length
    for (let i = 1; i < maxVariants; i++) {
        const newLength = baseLength + i;
        if (newLength > maxPossibleLength) break;
        
        const variantParams = { ...params };
        variantParams.path_length = newLength;
        yield this.generatePathInfo(gridSize, variantParams);
    }
  }

  generatePathInfo(
    gridSize: [number, number, number],
    params: Record<string, any>
  ): IPathInfo {
    console.log("    LOG: Generating 'straight_line' topology...");

    let pathLength = params.path_length || 5;
    const maxDim = Math.max(gridSize[0], gridSize[2]);
    if (pathLength >= maxDim - 2) pathLength = maxDim - 3;

    // Random choice
    const axis = Math.random() < 0.5 ? 'x' : 'z';
    const direction = Math.random() < 0.5 ? 1 : -1;

    // Start Pos
    const totalCellsNeeded = pathLength + 2;
    const y = 0;
    const startPosList: [number, number, number] = [0, y, 0];

    if (axis === 'x') {
        if (direction === 1) {
            startPosList[0] = Math.floor(Math.random() * (gridSize[0] - totalCellsNeeded + 1));
        } else {
            startPosList[0] = Math.floor(Math.random() * (gridSize[0] - totalCellsNeeded) + totalCellsNeeded - 1); // Python: randint(needed-1, size-1)
            // Fix: randint(a, b) includes b. 
            // Range: [total_cells_needed - 1, grid_size[0] - 1]
            const min = totalCellsNeeded - 1;
            const max = gridSize[0] - 1;
            startPosList[0] = Math.floor(Math.random() * (max - min + 1)) + min;
        }
        startPosList[2] = Math.floor(Math.random() * gridSize[2]);
    } else { // z
        if (direction === 1) {
            startPosList[2] = Math.floor(Math.random() * (gridSize[2] - totalCellsNeeded + 1));
        } else {
            const min = totalCellsNeeded - 1;
            const max = gridSize[2] - 1;
            startPosList[2] = Math.floor(Math.random() * (max - min + 1)) + min;
        }
        startPosList[0] = Math.floor(Math.random() * gridSize[0]);
    }

    const startPos: Coord = [startPosList[0], startPosList[1], startPosList[2]];
    const pathCoords: Coord[] = [startPos];
    const currentPos: Coord = [...startPos];

    // Create coords
    for (let i = 0; i < pathLength; i++) {
        if (axis === 'x') currentPos[0] += direction;
        else currentPos[2] += direction;
        pathCoords.push([...currentPos]);
    }

    const targetPos = pathCoords[pathCoords.length - 1];
    const midpoint = pathCoords[Math.floor(pathCoords.length / 2)];

    // Metadata
    // Python has: "segment": path_coords (single list)
    // NO "segments" key.
    
    const metadata = {
        topology_type: "straight_line",
        segment: pathCoords,
        path_length: pathLength,
        axis: axis,
        direction: direction,
        semantic_positions: {
            start: startPos,
            end: targetPos,
            midpoint: midpoint,
            optimal_start: 'start',
            optimal_end: 'end',
            valid_pairs: [
                {
                    name: 'full_line_easy',
                    start: 'start',
                    end: 'end',
                    path_type: 'full_traversal',
                    strategies: ['alternating_patterns', 'progressive_spacing'],
                    difficulty: 'EASY',
                    teaching_goal: 'Regular spacing along line'
                },
                {
                    name: 'half_line_medium',
                    start: 'start',
                    end: 'midpoint',
                    path_type: 'partial_traversal',
                    strategies: ['alternating_patterns', 'grouped_items'],
                    difficulty: 'MEDIUM',
                    teaching_goal: 'Grouped pattern in half'
                },
                {
                    name: 'variable_spacing_hard',
                    start: 'start',
                    end: 'end',
                    path_type: 'variable_spacing',
                    strategies: ['alternating_patterns', 'progressive_spacing'],
                    difficulty: 'HARD',
                    teaching_goal: 'Progressive item spacing'
                }
            ]
        }
    };

    return {
        start_pos: startPos,
        target_pos: targetPos,
        path_coords: pathCoords,
        placement_coords: pathCoords,
        obstacles: [],
        metadata: metadata
    };
  }
}
